Using different types of aggregates (#836)
mozilla/data-docs: src/cookbooks/data_modeling/aggregates_guidelines.md

# Introduction

The best approach to improve performance and reduce cost is to reduce the amount of data scanned by pre-calculating metrics and summarizing big datasets. This is where aggregates are most useful and flexible depending on the use case.

This doc discusses three options to aggregate data and provides a guideline of when to use each of them, with links to more detailed documentation and examples.

> 1. BigQuery aggregates tables. There are tables that can contain some business logic and do not contain client_id. They are created and maintained by the Data Team and scheduled in Airflow.

Reviewer: ```suggestion
1. **BigQuery aggregate tables.** Aggregates tables contain pre-aggregated results of scheduled queries running some business logic. These tables are created and maintained by the Data Team and scheduled to get updated periodically via Airflow.
```

  [Cookbook link to create an aggregate table](https://mozilla.github.io/bigquery-etl/cookbooks/common_workflows/#adding-a-new-scheduled-query).

> 1. BigQuery Materialized Views. These are views defined by the developer using only SELECT and created, managed and scheduled by BigQuery. BigQuery creates aggregate tables to store the results of the view definition and thus optimize the queries.
>   [TODO: Instructions to implement a Materialized view]

Reviewer: ```suggestion
1. **[BigQuery Materialized Views](https://cloud.google.com/bigquery/docs/materialized-views-intro).** These are views defined by the developer using only SELECT and created, managed and scheduled by BigQuery. Materialized views are periodically updated reading only changes from the based table to compute results. Materialized view definitions [*do not support* certain BigQuery features and expressions](https://cloud.google.com/bigquery/docs/materialized-views-intro#limitations), such as UDFs, certain aggregate functions, backfilling or nesting.
```

>   ```
>   CREATE MATERIALIZED VIEW IF NOT EXISTS <dataset>.<materialized_view_name>
>   PARTITION BY
>     [partition_fields]
>   CLUSTER BY
>     [cluster_fields]
>   OPTIONS
>     (enable_refresh = TRUE, refresh_interval_minutes = 720)
>   AS
>   WITH <subquery_name> AS ( <subquery> ),
>   SELECT
>     <[columns]>
>   FROM
>     <subquery_name>
>   WHERE
>     <partition_field> > TIMESTAMP(<starting_date_to_backfill_the_view>)
>   GROUP BY
>     <[fields]>
>   ```

Reviewer: ```suggestion
CREATE MATERIALIZED VIEW `moz-fx-data-shared-prod.monitoring_derived.suggest_click_rate_live_v1`
OPTIONS
  (enable_refresh = TRUE, refresh_interval_minutes = 5)
AS
SELECT
  TIMESTAMP_TRUNC(submission_timestamp, minute) AS submission_minute,
  COUNT(*) AS n,
  COUNTIF(release_channel = "release") AS n_release,
  COUNTIF(release_channel = "beta") AS n_beta,
  COUNTIF(release_channel = "nightly") AS n_nightly,
  COUNT(request_id) AS n_merino,
  COUNTIF(request_id IS NOT NULL AND release_channel = "release") AS n_merino_release,
  COUNTIF(request_id IS NOT NULL AND release_channel = "beta") AS n_merino_beta,
  COUNTIF(request_id IS NOT NULL AND release_channel = "nightly") AS n_merino_nightly,
FROM
  `moz-fx-data-shared-prod.contextual_services_live.quicksuggest_click_v1`
WHERE
  DATE(submission_timestamp) > '2010-01-01'
GROUP BY
  1
```

>   - An example of a materialized view definition.

Author: @scholtzan do you have an example you could recommend?

>     1. Looker PDTs & aggregate awareness. These are aggregations that a developer defines in an Explore file (explore.lkml). From this definition, Looker creates a table in BigQuery's `mozdata.temp` using the naming convention`scratch schema + table status code + hash value + view name` and runs the scheduled update of the data.

Reviewer: ```suggestion
    1. **Looker PDTs & aggregate awareness.** These are aggregations that a developer defines in an Explore file (`explore.lkml`). From this definition, Looker creates a table in BigQuery's `mozdata.tmp` using the naming convention`scratch schema + table status code + hash value + view name` and runs the scheduled update of the data.
```

       ```
       aggregate_table: <aggregate_name: Descriptive name of this aggregation.> {
        query: {
        dimensions: [<table>.<columns>]
        measures: [<table>.<columns>]
        filters: [<table>.<partition_column>: "<Period to materialize in the aggregate e.g. 2 years>"]
        }
        materialization: {
        sql_trigger_value: SELECT CURRENT_DATE() ;;
        increment_key: <table>.<partition_column>
        increment_offset: <INT: number of periods to update, recommended is 1.> }
        }
       ```
       - An [Example of aggregate awareness in a Looker Explore](https://mozilla.cloud.looker.com/projects/spoke-default/files/combined_browser_metrics/explores/active_users_aggregates.explore.lkml)

## Important considerations:

- [Store the business logic in BigQuery](https://docs.telemetry.mozilla.org/cookbooks/data_modeling/where_to_store), ideally in a client_id level table to aggregate from.
- All aggregates are version controlled in the git repositories `bigquery-etl` and `spoke-default`.
- All types of aggregates require a backfill when the source data changes.
  - The managed-backfill process can be used for aggregate tables.

>   - The materialized views are backfilled by ?. Any schema change in the base table invalidates the view and requires backfill.

Author: @scholtzan I'm looking for the documentatino of how materialized views are currently backfilled for changes in previous data.

>   - The materialized views are backfilled by ?. Any schema change in the base table invalidates the view and requires backfill.

Reviewer: ```suggestion
  - Materialized views cannot be backfilled, instead a materialized view needs to be re-created. Schema changes in base tables invalidates the view and requires it to be re-created. Materialized views scan all historical data of their referenced base tables by default, so **ensure to set a date filter to reduce the amount of data to be scanned**.
  - Add a date filter to materialized views to limit the amount of data to be scanned when these views get deployed initially. Otherwise they will scan the entire data in referenced base tables.
```

  - Looker PDTs require following a [protocol to backfill](https://mozilla.udemy.com/course/looker-training-for-developers/learn/lecture/35440216/), currently not automated.
- Using indices, partition and clustering is allowed for all cases. Looker requires that they are defined in the base table.
- There is a limit of 20 materialized views per table.

> 

Reviewer: ```suggestion
- When considering materialized views, a common practice is to use a combination of an aggregate table to store historical data (e.g. older than 2 days) and use a materialized view to track data in real-time (e.g. aggregate data that is just coming in). This allows to run backfills on the aggregate table.
```

## What are the use cases to use each of these aggregates?

#### Use a BigQuery aggregate table when:

> - The results should not be unexpectedly affected by shredded.

Reviewer: ```suggestion
- The results should not be unexpectedly affected by Shredder.
```

- The metric requires strict change control.
- A visible alerts is required in case of failure or data out of sync. In this option, the scheduler sends email and alerts attended daily by the Data Engineering team during the Airflow Triage process.
- The table will be queried directly or used as a source for other analysis. Looker PDTs are not designed to be queried directly.
- The query requires full flexibility to use DML, data types, aggregation functions and different types of JOIN.

#### Use a Materialized View when:

> - The results can be based on shredded data (tables with client_id).

Reviewer: ```suggestion
- To aggregate data in real-time (for example, for implementing real-time monitoring of certain metrics)
- The results can be based on shredded data (tables with client_id).
```

- The view will be queried directly or used as a source for other analysis.
- Change control is **not** required or is already implemented in the base table. This can be verified by looking for the label `change_controlled: true` in the table's metadata.

> - An explicit alert on failure is **not** required. Failures need to be intentionally monitored in the Dashboard ?

Author: @scholtzan I'm looking for the Materialized monitoring dashboard here

- The metric does **not** require non-deterministic functions that are not supported: RAND(), CURRENT_TIMESTAMP, CURRENT_DATE(), or CURRENT_TIME().
- The query does **not** require UDFs, UNNESTING arrays, COUNT DISTINCT, ORDER BY or any DML operation different from SELECT.
- The query uses a WITH clause, COUNTIF, INNER JOIN or TIMESTAMP_ADD. These are all supported.

> 

Reviewer: ```suggestion
- The data does not need to be backfilled

```

#### Use a Looker PDT when:
- The results can be based on shredded data (tables with client_id).
- Change control is not required or is already implemented in the base table. This can be verified by looking for the label `change_controlled: true` in the metadata.

> - An explicit alert on failure is **not** required. Failures are monitored in the Dashboard [Errors and Broken Content Dashboard](https://mozilla.cloud.looker.com/dashboards/system__activity::errors_and_broken_content) or in the [PDTs Dashboard](https://mozilla.cloud.looker.com/admin/pdts).

Author: @scholtzan what do you think about adding this dashboard to the Airflow Triage, since there are many PDT failures that don't generate an alert.

Reviewer: Yeah, that makes sense. We could add a link to the health dashboard

Author: Sounds good, I added the link to PDTs admin to the [Platform Health Check](https://mozilla.cloud.looker.com/dashboards/387) so these are checked once a week.
